llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp

```diff
@@ -1408,7 +1408,20 @@ class AMDGPUAsmParser : public MCTargetAsmParser {
       copySTI().ToggleFeature("southern-islands");
     }
 
-    setAvailableFeatures(ComputeAvailableFeatures(getFeatureBits()));
+    StringRef FS = getSTI().getFeatureString();
+    FeatureBitset FB = getFeatureBits();
+    int isWave32 = FS.contains("+wavefrontsize32");
+    int isWave64 = FS.contains("+wavefrontsize64");
+    // Reset default wavefrontsize features.
+    if (isWave32 + isWave64 == 1) {
+      if (isWave32 && FB[AMDGPU::FeatureWavefrontSize64])
```

Comment:
We shouldn't really be directly checking the bits. The + and -s are sequentially evaluated in the subtarget construction to set the WavefrontSize field. I would expect the same to happen down in MC. I don't know why the subtarget evaluation logic isn't an an MCSubtarget superclass

Reply:
We do exactly the same in the codegen: https://llvm.org/doxygen/AMDGPUSubtarget_8cpp_source.html#l00085

Reply:
That's also hacky pre-processing of the input features strings, which should also be avoidable. The actual interpretation logic is in ParseSubtargetFeatures

Reply:
We can remove FeatureWavefrontSize32 from gfx10 and above, but then we need to add default wavefront size feature everywhere, and we need to parse feature string for this anyway. We are already doing it in clang and flang, but not in any other tools.

Reply:
Like this: https://github.com/llvm/llvm-project/pull/98400. Actually w/o additional string parsing.